### PR TITLE
Ph

### PR DESCRIPTION
**Pull request by olliethomas**

### CODE REVIEW

**Comment by olliethomas** [comment]
Still got work to do, but I think it maybe makes more sense to expect the classifier to be an `nn.Module`

File: bolts/callbacks/posthoc_eval.py [hunk 1]
```diff
@@ -1,18 +1,63 @@
...
+        with torch.no_grad():
+            self.enc.eval()
+            z = self.enc(x)
+        return self.clf(z)
+
+    def test_step(self, batch: NamedTuple, batch_idx: int) -> Optional[STEP_OUTPUT]:
+        logits = self(batch.x)
+        return {
+            "y": batch.y.view(-1),
+            "preds": logits.sigmoid().round().squeeze(-1),
```
Comment (MylesBartlett) [inline bolts/callbacks/posthoc_eval.py:36]: Can we have this be generalizable?

File: bolts/callbacks/posthoc_eval.py [hunk 2]
```diff
@@ -1,18 +1,63 @@
...
+    def forward(self, x: Tensor) -> Tensor:
+        with torch.no_grad():
+            self.enc.eval()
+            z = self.enc(x)
+        return self.clf(z)
+
+    def test_step(self, batch: NamedTuple, batch_idx: int) -> Optional[STEP_OUTPUT]:
+        logits = self(batch.x)
+        return {
+            "y": batch.y.view(-1),
```
Comment (MylesBartlett) [inline bolts/callbacks/posthoc_eval.py:35]:
```suggestion
            "y": batch.y.flatten(),
```

File: bolts/callbacks/posthoc_eval.py [hunk 3]
```diff
@@ -1,18 +1,63 @@
...
+class EvalModule(pl.LightningModule):
+    def __init__(self, enc: nn.Module, clf: nn.Module):
+        super().__init__()
+        self.enc = enc
+        self.clf = gcopy(clf)
+        self._loss = nn.CrossEntropyLoss()
+
+    def training_step(self, batch: NamedTuple, batch_idx: int) -> STEP_OUTPUT:
+        y = self(batch.x)
+        return self._loss(y, batch.y.view(-1).long())
```
Comment (MylesBartlett) [inline bolts/callbacks/posthoc_eval.py:24]:
```suggestion
        return self._loss(y, batch.y.flatten().long())
```

File: bolts/callbacks/posthoc_eval.py [hunk 4]
```diff
@@ -1,18 +1,63 @@
...
+    def test_epoch_end(self, outputs: EPOCH_OUTPUT) -> None:
+        torch.cat([_r["y"] for _r in outputs], 0)
+        torch.cat([_r["preds"] for _r in outputs], 0)
+
+    @staticmethod
+    def _maybe_reset_parameters(module: nn.Module) -> None:
+        if hasattr(module, 'reset_parameters'):
+            module.reset_parameters()  # type: ignore
+
+    def reset_parameters(self):
```
Comment (MylesBartlett) [inline bolts/callbacks/posthoc_eval.py:48]:
```suggestion
    def reset_parameters(self) -> None:
```

File: bolts/callbacks/posthoc_eval.py [hunk 5]
```diff
@@ -1,18 +1,63 @@
...
+        self.clf.apply(self._maybe_reset_parameters)
+
+    def configure_optimizers(self):
```
Comment (MylesBartlett) [inline bolts/callbacks/posthoc_eval.py:51]:
```suggestion
    def configure_optimizers(self) -> Optimizer:
```

File: bolts/callbacks/posthoc_eval.py [hunk 6]
```diff
@@ -1,18 +1,63 @@
...
+import torch
+from torch import Tensor, nn
+from torch.optim import AdamW
 from torch.utils.data import DataLoader
 
 __all__ = ["PostHocEval"]
 
 
+class EvalModule(pl.LightningModule):
+    def __init__(self, enc: nn.Module, clf: nn.Module):
```
Comment (MylesBartlett) [inline bolts/callbacks/posthoc_eval.py:16]:
```suggestion
    def __init__(self, enc: nn.Module, clf: nn.Module) -> None:
```

File: tests/callbacks_test.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import pytorch_lightning as pl
+from torch import nn
+
+from bolts.callbacks import PostHocEval
+from bolts.fair.models import ErmBaseline
+from tests.fair.model_test import DummyDataModule, Encoder  # type: ignore[import]
```
Comment (thomkeh) [inline tests/callbacks_test.py:6]: why can't mypy import this?

Comment (olliethomas) [reply tests/callbacks_test.py:6]: because `tests` is a dir not a module

Comment (olliethomas) [reply tests/callbacks_test.py:6]: i read somewhere that this is a preferred way of structuring tests, but now I can't find it, and can only find sources that say _do_ include `__init__.py`

Comment (olliethomas) [reply tests/callbacks_test.py:6]: I added __init__.py files... still can't find them -_-

Comment (olliethomas) [reply tests/callbacks_test.py:6]: ah, was missing one

File: bolts/callbacks/posthoc_eval.py [hunk 8]
```diff
@@ -1,18 +1,78 @@
...
+import torch
+from torch import Tensor, nn
+from torch.optim import AdamW
 from torch.utils.data import DataLoader
+import torchmetrics
 
 __all__ = ["PostHocEval"]
 
 
+class EvalTuple(NamedTuple):
```
Comment (MylesBartlett) [inline bolts/callbacks/posthoc_eval.py:16]: Use the newly-added BinaryTuple? This will require a merge with master.

Comment (olliethomas) [reply bolts/callbacks/posthoc_eval.py:16]: Where is BinaryTuple?

Comment (MylesBartlett) [reply bolts/callbacks/posthoc_eval.py:16]: BinarySample

**Review by MylesBartlett — APPROVED** [review]

**Pull request merged**